File: python/server/src/smooth_operator_server/postgres_knowledge.py
"""Durable, vector-searched knowledge storage for the Python server — the sibling of
the Rust ``rust/adapters/postgres/src/knowledge.rs`` and the .NET
``PostgresKnowledgeBase`` / ``PostgresAclKnowledgeStore``.

Two stores, both on Postgres + ``pgvector``, both over the SAME ``knowledge_vectors``
table the Rust adapter defines (``rust/adapters/postgres/src/schema.rs``) so every
server in this repo shares one set of tables — same names, same columns:

- :class:`PostgresVectorKnowledge` — ingest documents, retrieve by cosine similarity.
  The durable analog of core's in-memory :class:`~smooth_operator_core.VectorKnowledge`.
- :class:`PostgresAclKnowledge` — the same, plus document-level access control. Each
  document carries an ACL (``{public, users, groups}``) persisted in the ``acl`` jsonb
  column; :meth:`~PostgresAclKnowledge.for_access` returns a read-only view that filters
  by the caller's entitlements **in SQL** — a restricted document is never even fetched.
  Mirrors Rust's ``PgKnowledgeBase::with_access`` and .NET's ``ForAccess``.

Retrieval reuses the offline, deterministic :class:`~smooth_operator_core.HashEmbedder`
by default (no network, stable across processes), sizing the ``vector(N)`` column to the
embedder's dimension. A gateway embedder drops in behind the same ``embed(text)`` seam.

``asyncpg`` is an OPTIONAL dependency (the ``postgres`` extra); importing this module is
what pulls it in. The embedding is stored by casting a pgvector text literal
(``$n::text::vector``) so no asyncpg type codec has to be registered — the same trick
the Rust adapter uses.
"""

from __future__ import annotations

import json
import uuid
from dataclasses import dataclass, field
from typing import Any, Optional, Protocol

import asyncpg
from smooth_operator_core import HashEmbedder, KnowledgeHit


class Embedder(Protocol):
    """Turns text into a fixed-length vector (core's :class:`~smooth_operator_core.Embedder`)."""

    def embed(self, text: str) -> list[float]: ...


# ── ACL value types ──────────────────────────────────────────────────────────
# Minimal mirrors of the Rust ``DocAcl`` / ``AccessContext`` and .NET ``DocumentAcl`` /
# ``AccessContext``. Core has no ACL types (the in-memory stores are ACL-free), so they
# live here with the durable store that needs them.


@dataclass(frozen=True)
class DocumentAcl:
    """A document's access control: public, or restricted to named users / groups.

    Serialized to the ``acl`` jsonb column as ``{public, users, groups}`` — the same
    shape Rust's ``DocAcl`` persists, so a row written here is readable cross-server."""

    public: bool = False
    users: tuple[str, ...] = ()
    groups: tuple[str, ...] = ()

    @classmethod
    def public_acl(cls) -> "DocumentAcl":
        return cls(public=True)

    @classmethod
    def for_groups(cls, *groups: str) -> "DocumentAcl":
        return cls(public=False, groups=tuple(groups))

    def to_json(self) -> str:
        return json.dumps({"public": self.public, "users": list(self.users), "groups": list(self.groups)})


@dataclass(frozen=True)
class AccessContext:
    """A requester's entitlements. ``anonymous`` carries no groups → public docs only."""

    user_id: Optional[str] = None
    groups: tuple[str, ...] = field(default=())
    anonymous: bool = True

    #: The anonymous requester — sees only public documents (fail-closed).
    @classmethod
    def anon(cls) -> "AccessContext":
        return cls(anonymous=True)

    @classmethod
    def for_groups(cls, *groups: str, user_id: str | None = None) -> "AccessContext":
        return cls(user_id=user_id, groups=tuple(groups), anonymous=len(groups) == 0)


def _vector_literal(vec: list[float]) -> str:
    """A pgvector text literal: ``[0.1,0.2,...]``. Cast ``::text::vector`` in SQL."""
    return "[" + ",".join(str(x) for x in vec) + "]"


# The ``knowledge_vectors`` DDL, verbatim in spirit from the Rust adapter's
# ``knowledge_vectors_schema(dim)`` so all servers share one table. The dimension is the
# active embedder's, so the column width always matches the vectors written into it.
def _schema(dim: int) -> str:
    return f"""
CREATE EXTENSION IF NOT EXISTS vector;
CREATE TABLE IF NOT EXISTS knowledge_vectors (
    id              TEXT PRIMARY KEY,
    document_id     TEXT NOT NULL,
    organization_id TEXT,
    source          TEXT NOT NULL,
    content         TEXT NOT NULL,
    embedding       vector({dim}) NOT NULL,
    content_tsv     tsvector GENERATED ALWAYS AS (to_tsvector('english', content)) STORED,
    metadata        JSONB,
    acl             JSONB,
    created_at      TIMESTAMPTZ NOT NULL DEFAULT now()
);
-- Idempotent add so a table created before the ACL column existed gains it in place.
ALTER TABLE knowledge_vectors ADD COLUMN IF NOT EXISTS acl JSONB;
CREATE INDEX IF NOT EXISTS idx_knowledge_content_tsv
    ON knowledge_vectors USING gin (content_tsv);
CREATE INDEX IF NOT EXISTS idx_knowledge_org
    ON knowledge_vectors (organization_id);
"""


class PostgresVectorKnowledge:
    """Durable vector knowledge on Postgres + pgvector. Cosine-ranked retrieval.

    Ingest is idempotent by document id (a document is one chunk keyed by its id, so
    re-ingesting the same id upserts in place). Retrieval embeds the query and ranks by
    pgvector cosine distance (``<=>``), returning core :class:`KnowledgeHit`\\ s so the
    result is protocol-compatible with the in-memory retrievers."""

    def __init__(self, pool: asyncpg.Pool, embedder: Embedder) -> None:
        self._pool = pool
        self._embedder = embedder

    @classmethod
    async def create(cls, dsn: str, embedder: Embedder | None = None) -> "PostgresVectorKnowledge":
        """Connect, create the pgvector extension + ``knowledge_vectors`` table, and
        return the store. The vector column is sized to the embedder's dimension,
        discovered by embedding a probe string so any :class:`Embedder` works."""
        embedder = embedder or HashEmbedder()
        dim = len(embedder.embed("dimension probe"))
        pool = await asyncpg.create_pool(dsn)
        assert pool is not None
        try:
            await pool.execute(_schema(dim))
        except BaseException:
            await pool.close()
            raise
        return cls(pool, embedder)

    async def close(self) -> None:
        await self._pool.close()

    async def ingest(
        self,
        content: str,
        source: str,
        *,
        document_id: str | None = None,
        organization_id: str | None = None,
        metadata: dict[str, Any] | None = None,
        acl: DocumentAcl | None = None,
    ) -> str:
        """Embed and upsert a document, returning its id. Idempotent by id: passing the
        same ``document_id`` again refreshes the row rather than duplicating it."""
        doc_id = document_id or str(uuid.uuid4())
        embedding = _vector_literal(self._embedder.embed(content))
        await self._pool.execute(
            """INSERT INTO knowledge_vectors
                   (id, document_id, organization_id, source, content, embedding, metadata, acl)
               VALUES ($1, $1, $2, $3, $4, $5::text::vector, $6::jsonb, $7::jsonb)
               ON CONFLICT (id) DO UPDATE SET
                   organization_id = EXCLUDED.organization_id,
                   source          = EXCLUDED.source,
                   content         = EXCLUDED.content,
                   embedding       = EXCLUDED.embedding,
                   metadata        = EXCLUDED.metadata,
                   acl             = EXCLUDED.acl""",
            doc_id,
            organization_id,
            source,
            content,
            embedding,
            json.dumps(metadata) if metadata is not None else None,
            acl.to_json() if acl is not None else None,
        )
        return doc_id

    async def query(self, query: str, top_k: int = 4, *, organization_id: str | None = None) -> list[KnowledgeHit]:
        """The ``top_k`` documents nearest the query by cosine similarity. ``score`` is
        ``1 - cosine_distance`` (higher is closer), matching the .NET store."""
        if top_k <= 0:
            return []
        rows = await self._query_rows(query, top_k, organization_id=organization_id, access=None)
        return [KnowledgeHit(content=r["content"], source=r["source"], score=float(r["score"])) for r in rows]

    async def _query_rows(
        self,
        query: str,
        top_k: int,
        *,
        organization_id: str | None,
        access: AccessContext | None,
    ) -> list[asyncpg.Record]:
        literal = _vector_literal(self._embedder.embed(query))
        # ACL predicate (only when this is an access-scoped read). A row is visible when
        # it has no recorded ACL (org-public default), is explicitly public, names the
        # requester's user id, or names any of the requester's groups — mirroring the
        # Rust ``query_async`` filter. ``?`` / ``?|`` are jsonb key-exists operators.
        if access is None:
            return await self._pool.fetch(
                """SELECT content, source, 1 - (embedding <=> $1::text::vector) AS score
                     FROM knowledge_vectors
                    WHERE ($2::text IS NULL OR organization_id = $2)
                    ORDER BY embedding <=> $1::text::vector
                    LIMIT $3""",
                literal,
                organization_id,
                top_k,
            )
        return await self._pool.fetch(
            """SELECT content, source, 1 - (embedding <=> $1::text::vector) AS score
                 FROM knowledge_vectors
                WHERE ($2::text IS NULL OR organization_id = $2)
                  AND (acl IS NULL
                       OR (acl->>'public')::boolean IS TRUE
                       OR ($4::text IS NOT NULL AND acl->'users' ? $4)
                       OR (acl->'groups' ?| $5::text[]))
                ORDER BY embedding <=> $1::text::vector
                LIMIT $3""",
            literal,
            organization_id,
            top_k,
            access.user_id,
            list(access.groups),
        )


class _ScopedView:
    """A read-only knowledge view bound to one requester's entitlements — the object
    returned by :meth:`PostgresAclKnowledge.for_access`. Its :meth:`query` filters by the
    ACL in SQL; it cannot ingest (mirrors .NET's ``ScopedView``)."""

    def __init__(self, store: "PostgresAclKnowledge", access: AccessContext) -> None:
        self._store = store
        self._access = access

    async def query(self, query: str, top_k: int = 4, *, organization_id: str | None = None) -> list[KnowledgeHit]:
        if top_k <= 0:
            return []
        rows = await self._store._backing._query_rows(
            query, top_k, organization_id=organization_id, access=self._access
        )
        return [KnowledgeHit(content=r["content"], source=r["source"], score=float(r["score"])) for r in rows]


class PostgresAclKnowledge:
    """Durable, ACL-aware vector knowledge. Ingest carries a :class:`DocumentAcl`;
    :meth:`for_access` returns a view whose queries a restricted document never reaches.
    Same ``knowledge_vectors`` table and cosine ranking as
    :class:`PostgresVectorKnowledge` — the ACL is one jsonb column and one SQL predicate.
    """

    def __init__(self, backing: PostgresVectorKnowledge) -> None:
        self._backing = backing

    @classmethod
    async def create(cls, dsn: str, embedder: Embedder | None = None) -> "PostgresAclKnowledge":
        return cls(await PostgresVectorKnowledge.create(dsn, embedder))

    async def close(self) -> None:
        await self._backing.close()

    async def ingest(
        self,
        content: str,
        source: str,
        acl: DocumentAcl,
        *,
        document_id: str | None = None,
        organization_id: str | None = None,
    ) -> str:
        return await self._backing.ingest(
            content, source, document_id=document_id, organization_id=organization_id, acl=acl
        )

    def for_access(self, access: AccessContext) -> _ScopedView:
        """A read-only view enforcing ``access``'s entitlements in SQL."""
        return _ScopedView(self, access)

File: typescript/server/src/postgresKnowledge.ts
/**
 * Durable, vector-searched knowledge storage for the TypeScript server — the TS
 * sibling of the Rust `PgKnowledgeBase` (`rust/adapters/postgres/src/knowledge.rs`)
 * and the C# `PostgresKnowledgeBase` / `PostgresAclKnowledgeStore`
 * (`dotnet/server/postgres/src`).
 *
 * A document is embedded (via an injected {@link Embedder}) and stored as one row
 * in the shared `knowledge_vectors` table — the SAME table the Rust adapter creates
 * (`rust/adapters/postgres/src/schema.rs`), so a row written here reads back in every
 * other server in this repo. Retrieval embeds the query and ranks by pgvector cosine
 * distance (`embedding <=> $q`). Access control lives in the `acl` JSONB column and is
 * filtered IN SQL — a restricted document is never even fetched — exactly as the Rust
 * `query_async` ACL predicate does.
 *
 * ponytail: cosine-only dense retrieval. Rust additionally runs a sparse tsvector arm
 * and fuses the two with RRF; the shared behavioural contract (ingest → retrieve the
 * relevant doc, idempotent by id, ACL leak boundary) is satisfied by dense alone (the
 * C# reference is dense-only too). The `content_tsv` column still comes free with the
 * shared schema. Add the sparse arm + RRF here if lexical recall proves it earns its keep.
 */

import { Pool } from 'pg';

import type { AccessContext } from './auth.js';

/** A document to ingest. Mirrors the C# `KnowledgeDocument` / Rust `Document`. */
export interface KnowledgeDocument {
    id: string;
    content: string;
    source: string;
}

/** A retrieval hit. Mirrors the C# `KnowledgeResult` (documentId + chunk + score + source). */
export interface KnowledgeQueryResult {
    documentId: string;
    chunk: string;
    score: number;
    source: string;
}

/**
 * Document-level access control, persisted as the `acl` JSONB column. The serialized
 * shape (`{public, users, groups}`) is byte-compatible with the Rust `DocAcl`, so an
 * ACL written here filters correctly in the Rust adapter and vice-versa. NULL column
 * ⇒ no ACL recorded ⇒ org-public (the backward-compatible default).
 */
export interface DocumentAcl {
    public: boolean;
    users?: string[];
    groups?: string[];
}

/** A public document — visible to everyone. Mirrors C# `DocumentAcl.PublicAcl`. */
export const PUBLIC_ACL: DocumentAcl = { public: true };

/** An ACL restricting a document to the given entitlement groups. Mirrors `DocumentAcl.ForGroups`. */
export function aclForGroups(...groups: string[]): DocumentAcl {
    return { public: false, groups };
}

/**
 * Turns text into an embedding vector for similarity search. The seam mirrors the
 * Rust/C# `Embedder`: a real gateway embedder for production, a deterministic one for
 * tests / offline use.
 */
export interface Embedder {
    readonly dimensions: number;
    embed(text: string): Promise<number[]>;
}

/**
 * A deterministic, network-free embedder — hashed bag-of-words into a fixed-dimension,
 * L2-normalized vector. Same text → same vector; texts sharing tokens are close in
 * cosine space. The TS analog of the Rust/C# `DeterministicEmbedder`; ideal for tests
 * and small in-process corpora.
 */
export class DeterministicEmbedder implements Embedder {
    readonly dimensions: number;

    constructor(dimensions = 256) {
        if (dimensions <= 0) throw new RangeError('dimensions must be positive');
        this.dimensions = dimensions;
    }

    // eslint-disable-next-line @typescript-eslint/require-await -- async to satisfy the Embedder seam
    async embed(text: string): Promise<number[]> {
        const vector = new Array<number>(this.dimensions).fill(0);
        for (const token of tokenize(text)) {
            const slot = fnv1a(token) % this.dimensions;
            vector[slot] = (vector[slot] ?? 0) + 1;
        }
        // L2-normalize so cosine distance is well-behaved.
        let sumOfSquares = 0;
        for (const value of vector) sumOfSquares += value * value;
        if (sumOfSquares > 0) {
            const norm = Math.sqrt(sumOfSquares);
            for (let i = 0; i < vector.length; i++) vector[i] = (vector[i] ?? 0) / norm;
        }
        return vector;
    }
}

/** Lowercase, strip non-alphanumerics, keep tokens longer than 2 chars. Mirrors the C# tokenizer. */
function tokenize(text: string): string[] {
    const tokens: string[] = [];
    for (const raw of text.toLowerCase().split(/\s+/)) {
        const token = raw.replace(/[^a-z0-9]/g, '');
        if (token.length > 2) tokens.push(token);
    }
    return tokens;
}

/** 32-bit FNV-1a, unsigned. Same constants as the Rust/C# reference. */
function fnv1a(value: string): number {
    let hash = 2166136261;
    for (let i = 0; i < value.length; i++) {
        hash ^= value.charCodeAt(i);
        hash = Math.imul(hash, 16777619);
    }
    return hash >>> 0;
}

/** Format a vector as a pgvector literal: `[0.1,0.2,...]`. */
function vectorLiteral(v: number[]): string {
    return `[${v.join(',')}]`;
}

/**
 * The DDL applied on connect — the `knowledge_vectors` table verbatim from the Rust
 * adapter (`rust/adapters/postgres/src/schema.rs`, `knowledge_vectors_schema`), so the
 * table is byte-identical across every server. Requires a pgvector-enabled Postgres
 * (`pgvector/pgvector:pg16`). Every statement is idempotent.
 */
function schema(dim: number): string {
    return `
CREATE EXTENSION IF NOT EXISTS vector;
CREATE TABLE IF NOT EXISTS knowledge_vectors (
    id              TEXT PRIMARY KEY,
    document_id     TEXT NOT NULL,
    organization_id TEXT,
    source          TEXT NOT NULL,
    content         TEXT NOT NULL,
    embedding       vector(${dim}) NOT NULL,
    content_tsv     tsvector GENERATED ALWAYS AS (to_tsvector('english', content)) STORED,
    metadata        JSONB,
    acl             JSONB,
    created_at      TIMESTAMPTZ NOT NULL DEFAULT now()
);
ALTER TABLE knowledge_vectors ADD COLUMN IF NOT EXISTS acl JSONB;
CREATE INDEX IF NOT EXISTS idx_knowledge_embedding_hnsw
    ON knowledge_vectors USING hnsw (embedding vector_cosine_ops);
CREATE INDEX IF NOT EXISTS idx_knowledge_content_tsv
    ON knowledge_vectors USING gin (content_tsv);
CREATE INDEX IF NOT EXISTS idx_knowledge_org
    ON knowledge_vectors (organization_id);
`;
}

/**
 * The ACL visibility predicate, applied in SQL to every query. A row is visible when it
 * has no recorded ACL (org-public), is explicitly public, names the requester's user id,
 * or names any of the requester's groups (`?` / `?|` are jsonb key-exists operators).
 * Verbatim from the Rust `query_async` predicate. `$2` is the requester user id (NULL ⇒
 * anonymous), `$3` the requester groups (text[]).
 */
const ACL_PREDICATE = `(acl IS NULL
    OR (acl->>'public')::boolean IS TRUE
    OR ($2::text IS NOT NULL AND acl->'users' ? $2)
    OR (acl->'groups' ?| $3::text[]))`;

/**
 * A durable, vector-searched, ACL-aware knowledge store on one Postgres + pgvector pool.
 *
 * Follows the Rust model of ONE type carrying optional access, rather than the C# split
 * into two classes: the plain retrieval path is {@link query} with no access (only
 * ACL-null/public rows match), and the chat path is {@link forAccess}, whose queries
 * filter rows by the requester's entitlements in SQL. {@link withAcl} mirrors the C#
 * ingest view for writing a document with an ACL.
 */
export class PostgresKnowledgeStore {
    private constructor(
        private readonly pool: Pool,
        private readonly embedder: Embedder,
    ) {}

    /** Connect, enable pgvector, and apply the shared schema (all idempotent). */
    static async create(connectionString: string, embedder: Embedder): Promise<PostgresKnowledgeStore> {
        const pool = new Pool({ connectionString });
        try {
            await pool.query(schema(embedder.dimensions));
        } catch (error) {
            await pool.end();
            throw error;
        }
        return new PostgresKnowledgeStore(pool, embedder);
    }

    /** Release the connection pool. */
    async close(): Promise<void> {
        await this.pool.end();
    }

    /**
     * Insert or update a document, embedding its content. Upsert is keyed by document id
     * (the doc is stored as a single chunk keyed by its id, so re-ingesting the same doc
     * replaces it in place — idempotent by id). `acl` NULL ⇒ org-public.
     */
    async ingest(doc: KnowledgeDocument, acl?: DocumentAcl): Promise<void> {
        const embedding = await this.embedder.embed(doc.content);
        await this.pool.query(
            `INSERT INTO knowledge_vectors (id, document_id, source, content, embedding, acl)
             VALUES ($1, $1, $2, $3, $4::text::vector, $5::jsonb)
             ON CONFLICT (id) DO UPDATE SET
                 document_id = EXCLUDED.document_id,
                 source      = EXCLUDED.source,
                 content     = EXCLUDED.content,
                 embedding   = EXCLUDED.embedding,
                 acl         = EXCLUDED.acl`,
            [doc.id, doc.source, doc.content, vectorLiteral(embedding), acl ? JSON.stringify(acl) : null],
        );
    }

    /**
     * The top-`limit` documents by cosine similarity to `query`. Without an access
     * context only ACL-null/public rows match (the plain-retrieval contract); with one,
     * rows are filtered by the requester's user id + groups in SQL before ranking, so a
     * restricted document is never fetched. `score` is cosine similarity (`1 - distance`).
     */
    async query(query: string, limit: number, access?: AccessContext): Promise<KnowledgeQueryResult[]> {
        const embedding = await this.embedder.embed(query);
        const userId = access && !access.isAnonymous ? access.principal.sub : null;
        const groups = access ? access.principal.groups : [];
        const { rows } = await this.pool.query(
            `SELECT document_id, content, source, 1 - (embedding <=> $1::text::vector) AS score
               FROM knowledge_vectors
              WHERE ${ACL_PREDICATE}
              ORDER BY embedding <=> $1::text::vector
              LIMIT $4`,
            [vectorLiteral(embedding), userId, groups, limit > 0 ? limit : 0],
        );
        return rows.map((row) => ({
            documentId: row.document_id as string,
            chunk: row.content as string,
            score: Number(row.score),
            source: row.source as string,
        }));
    }

    /**
     * A read-only view whose {@link KnowledgeView.query} enforces `access`'s
     * entitlements. Mirrors the Rust `with_access` / C# `ForAccess`.
     */
    forAccess(access: AccessContext): KnowledgeView {
        return { query: (q, limit) => this.query(q, limit, access) };
    }

    /**
     * A write view whose {@link IngestView.ingest} stamps every document with `acl`.
     * Mirrors the C# `WithAcl`.
     */
    withAcl(acl: DocumentAcl): IngestView {
        return { ingest: (doc) => this.ingest(doc, acl) };
    }
}

/** A read-only, access-scoped retrieval view. */
export interface KnowledgeView {
    query(query: string, limit: number): Promise<KnowledgeQueryResult[]>;
}

/** A write view that stamps a fixed ACL onto every ingested document. */
export interface IngestView {
    ingest(doc: KnowledgeDocument): Promise<void>;
}
